src/pk.c

```diff
@@ -30,1707 +30,2849 @@
-    }
-    else if (outLen > *sigLen) {
-        WOLFSSL_MSG("Output buffer too small");
-        return WOLFSSL_FAILURE;
-    }
-    else if (wc_InitRng(tmpRNG) == 0) {
-        rng = tmpRNG;
-        initTmpRng = 1;
+        XFREE(rsa, heap, DYNAMIC_TYPE_RSA);
+        (void)heap;
     }
-    else {
-        WOLFSSL_MSG("Bad RNG Init, trying global");
+}
+
+/* Allocate and initialize a new RSA key.
+ *
+ * wolfSSL API.
+ *
+ * @param [in] heap   Heap hint.
+ * @param [in] devId  Device identifer value.
+ * @return  RSA key on success.
+ * @return  NULL on failure.
+ */
+WOLFSSL_RSA* wolfSSL_RSA_new_ex(void* heap, int devId)
+{
+    WOLFSSL_RSA* rsa = NULL;
+    RsaKey* key = NULL;
+    int err = 0;
+    int rsaKeyInited = 0;
+
+    WOLFSSL_ENTER("wolfSSL_RSA_new");
+
+    /* Allocate memory for new wolfCrypt RSA key. */
+    key = (RsaKey*)XMALLOC(sizeof(RsaKey), heap, DYNAMIC_TYPE_RSA);
+    if (key == NULL) {
+        WOLFSSL_MSG("wolfSSL_RSA_new malloc RsaKey failure");
+        err = 1;
+    }
+    if (!err) {
+        /* Allocate memory for new RSA key. */
+        rsa = (WOLFSSL_RSA*)XMALLOC(sizeof(WOLFSSL_RSA), heap,
+            DYNAMIC_TYPE_RSA);
+        if (rsa == NULL) {
+            WOLFSSL_MSG("wolfSSL_RSA_new malloc WOLFSSL_RSA failure");
+            err = 1;
+        }
+    }
+    if (!err) {
+        /* Clear all fields of RSA key. */
+        wolfssl_init_rsa(rsa);
+        /* Cache heap to use for all allocations. */
+        rsa->heap = heap;
+
+        /* Initialize reference counting. */
+        if (wolfSSL_RefInit(&rsa->ref) != 0) {
+            WOLFSSL_MSG("wc_InitMutex WOLFSSL_RSA failure");
+            err = 1;
+        }
+
+        /* Initialize wolfCrypt RSA key. */
+        if (wc_InitRsaKey_ex(key, heap, devId) != 0) {
+            WOLFSSL_MSG("InitRsaKey WOLFSSL_RSA failure");
+            err = 1;
+        }
+        else {
+            rsaKeyInited = 1;
+        }
+    }
+    #if !defined(HAVE_FIPS) && !defined(HAVE_USER_RSA) && \
+        !defined(HAVE_FAST_RSA) && defined(WC_RSA_BLINDING)
+    if (!err) {
+        WC_RNG* rng;
+
+        /* Create a local RNG. */
+        rng = (WC_RNG*)XMALLOC(sizeof(WC_RNG), heap, DYNAMIC_TYPE_RNG);
+        if ((rng != NULL) && (wc_InitRng_ex(rng, heap, devId) != 0)) {
+            WOLFSSL_MSG("InitRng failure, attempting to use global RNG");
+            XFREE(rng, heap, DYNAMIC_TYPE_RNG);
+            rng = NULL;
+        }
+
+        rsa->ownRng = 1;
+        if (rng == NULL) {
+            /* Get the wolfSSL global RNG - not thread safe. */
+            rng = wolfssl_get_global_rng();
+            rsa->ownRng = 0;
+        }
+        if (rng == NULL) {
+            /* Couldn't create global either. */
+            WOLFSSL_MSG("wolfSSL_RSA_new no WC_RNG for blinding");
+            err = 1;
+        }
+        else {
+            /* Set the local or global RNG into the wolfCrypt RSA key. */
+            (void)wc_RsaSetRNG(key, rng);
```

Comment:
Implementation is simply assigning the rng into the structure.
Previously the code didn't check the return but I'm explicitly not checking.
The error return is when key or rng are NULL.

Reply:
If `wc_RsaSetRNG` fails then `rng` will leak. Currently `wc_RsaSetRNG` will never fail since `key` and `rng` are NULL checked but we may want to expand this in the future.

Reply:
When implementation changes, all calls should be checked to ensure code will work.